### ISSUE TITLE
Unable to spin the motor from bash

### ISSUE THREAD

**Issue by bmegli** [bmegli/xv11lidar-test#1]
cxqzhou wrote:

Hi, I happened to find your article on how to use a LIDAR sensor on ev3dev.org and I became very interested in it, having a hobby for building robots. I followed your instructions on setting up the LIDAR sensor (I own the 5V version) for the EV3 here, but when I tried to follow your instructions on running your xv11lidar-test program, I ran into a few problems. Firstly, when I tried to start the motor through bash, it wouldn't start spinning, even though I know that the motor works since I got it to start spinning with a Python script. Here is screenshot of my terminal window to show you what I did: 

![t1](https://cloud.githubusercontent.com/assets/9095769/18901659/30cd1d3e-854d-11e6-9d05-46c27ecb8f75.png)

Secondly after failing to get the motor to start spinning, I tried to run your xv11lidar-test program, but the program did nothing, and I waited five minutes before terminating it. Here is another picture, where I tried to run your program:

![png2](https://cloud.githubusercontent.com/assets/9095769/18901667/4f1711b4-854d-11e6-93d1-f5bdd1cb663c.png)

So is there anything I can do to solve my issues? 
Thank you!

**Comment by bmegli** [comment]
Hi cxqzhou,

> Secondly after failing to get the motor to start spinning, I tried to run your xv11lidar-test program, but the program did nothing, and I waited five minutes before terminating

The LIDAR doesn't output any data if the motor is not spinning, `xv11test` on the other hand waits for the input to start synchronising so this behaviour is expected if the motor is not spinning. We have to solve the motor issue.

> I know that the motor works since I got it to start spinning with a Python script

Great - this means it's soldered correctly

It seems that you are loading `dc-motor` driver for a diffrent port from the one that the LIDAR motor is plugged in. It's possible to load this driver even for port that nothing is plugged in.

You can check the port with (in motor directory):

``` bash
cat address
```

Make sure it's the output that you have plugged in the motor.

Almost always the mapping is following:
- `port 4` -> `output A`
- `port 5` -> `output B`
- `port 6` -> `output C`
- `port 7` -> `output D`

But in theory this depends on the order the kernel loads the drivers of ports.

So summing up:

``` bash
echo dc-motor > /sys/class/lego-port/port4/mode #for output A, otherwise adjust the port number
dmesg | tail #you can see here which number the dc-motor  was created with
cat /sys/class/dc-motor/motorX/address # where X is the motor number, make sure the address matches the port
```

---

Now, there also other possibility.

Setting the port mode has to be done only once after boot.

I mean the line `echo dc-motor > /sys/class/lego-port/portX/mode` 

Considering that you have motor15 it means that you have changed the mode at least 16 times. It's possible that you are trying to set `duty_cycle_sp` for a different motor/port (again, dc-motor mode is kind of manual in ev3dev, nothing has to be plugged to port to set it and create dc-motor device)

So again, you can make sure you are setting `duty_cycle_sp` for the correct motor with just the same commands:

``` bash
echo dc-motor > /sys/class/lego-port/port4/mode #for output A, otherwise adjust the port number
dmesg | tail #you can see here which number the dc-motor  was created with
cat /sys/class/dc-motor/motorX/address # where X is the motor number, make sure the address matches the port the motor is plugged in
```

Let me know if this helps!

Kind regards

**Comment by cxqzhou8** [comment]
Hi @bmegli, Thanks for replying so quickly to my questions! I followed the your instructions above on setting the mode Output A port and now the LIDAR sensor's motor is spinning! (though I think the `duty_cycle_sp` should be set to -40 not 40) However when I tried to run the xv11test program again, it again did not do anything. Is there something else that I could possibly be doing wrong? Thanks again @bmegli!

**Comment by bmegli** [comment]
Hi @bobmcfoo,

> though I think the duty_cycle_sp should be set to -40 not 40

This depends how you solder the connector, but if you followowed the [tutorial](http://www.ev3dev.org/docs/tutorials/using-xv11-lidar/) it's 40 to rotate CCW. I have photographic documentation and 2 LIDAR units, one pair of cables soldered this week so I am pretty sure! ;-)

Whichever way you soldered doesn't really matter but **make it spin counterclockwise**, otherwise internal LIDAR speed calculation gets confused (kind of works but with a lot of errors).

> However when I tried to run the xv11test program again, it again did not do anything. Is there something else that I could possibly be doing wrong?

Make sure you have changed the input port mode to `other-uart`.  (this has to be done once, after you boot)

Almost always the mapping is the following:

- port 0 -> input 1 (/dev/tty_in1) 
- port 1 -> input 2 (/dev/tty_in2)
- port 2 -> input 3 (/dev/tty_in3)
- port 3 -> input 4 (/dev/tty_in4)

Again in theory this depends on the order the kernel loads the drivers of ports but I have not yet seen different order.

So:

*First* make the motor spin CCW, then:

```bash
echo other-uart > /sys/class/lego-port/port0/mode #for port 1
#wait a second, it may take kernel some time to react
./xv11test /dev/tty_in1
```
If you get nothing on the console then you may try:

## Increase duty cycle

Try increasing `duty_cycle_sp`, to say 45, the LIDAR doesn't output any data if its RPM is lower then 180. 

The effective speed the LIDAR is rotatating depends on:
- how much batteries are charged (voltage drops when discharging, the power that duty_cycle_sp gives depends on that
- on the unit (I have 2 units, both run with somewhat different speeds with the same duty_cycle)

## Test with picocom

xv11test will try to synchronize.

picocom will output anything there is on the line so:

```bash
apt-get install picocom
echo other-uart > /sys/class/lego-port/port0/mode # for input 1
picocom -b 115200 /dev/tty_in1
```

Then start spinning the LIDAR. If it is soldered correctly you should get message along the line:

`Spin! Press esc 3 times to cancel...' and then continous binary garbage like [here](https://www.youtube.com/watch?v=500PLf-LoOE)

If you get nothing then it's either not that port or there may be problem with soldering.

## Double check soldering

I tend to take a photo whenever I solder something before heat shrink tubing so that I can check if I have doubts. 

Make sure the cable colours match the ones in [tutorial](http://www.ev3dev.org/docs/tutorials/using-xv11-lidar/#data-and-power-connector) 

Specifically look at the photo, image 5 and 8 if you haven't soldered it the other way around (like possibly with motor if you have -40 instead of 40!)

----------------------------------------------------------------------------------------------

Nothing more comes to my mind at this moment.

**Comment by bmegli** [comment]
Ok @bobmcfoo , I have one more idea to troubleshoot.

The laser pulses are triggered by additional encoder trail on the LIDAR. 
It is enough for the LIDAR to spin to generate laser pulses, you don't have to read the data.

Those pulses are extremely short (200 microseconds) and are in the visible range of light.

## Safe Way

You need DSLR camera or other camera with adjustable exposure time.

Make the LIDAR spin, put it in a dark room. Set the camera for long exposure time (e.g. 1 minute) triggered after delay pointing the wall near the LIDAR.

When the process finishes you should have a photo of laser dots on the wall.

I did that in the past, with a friend from work, to test the angular accuracy of the LIDAR, I can't find the photo now (locally on some forgotten computer).

## Desperate Way

***AT YOUR OWN RISK!!!***

The laser is in the visible range of light. 
It's possible to see the laser dot (blinking shortly on every revolution) with naked eye if you look at exactly correct angle (on LIDAR level).

From theory, specification and Neato claims it is safe but those are claims and specs **for operation with original unit**.

I did it more than once. This doesn't mean that it is safe! Only that I believe it to be safe. Like everyody else I make errors all the time.

So if you do it this way, **do it at your own risk**, readup technical parameters of the unit, readup on safety classes and only then make informed decision.

## Summary

This will only tell you if the laser is powered, you may still need to troubleshot UART communication. 
Anyway, first try ideas from the first post.

## Additional Ideas

1. Have you plugged USB cable for LIDAR power?
2. Do you have means to measure the current consumption? LIDAR takes around 45 mA when idle and 135 mA when spinning, from USB. 
3. Do you use some exotic Wi-Fi dongle? I am using nano type that takes only 10 mA but I also have high power long-range Wi-Fi that takes 500 mA that would not probably work with the lidar from the passive USB hub.

**Comment by bmegli** [comment]
Hey, wait, we have solved the problem with the motor!

I am moving this to separate issue #2 for troubleshooting LIDAR data

So let's continue here #2

**Comment by bobmcfoo** [comment]
Thanks again for answering so quickly @bmegli! I seem to be having an issue installing picocom. 
```
 robot@ev3dev:~$ sudo apt-get install picocom
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following NEW packages will be installed:
  picocom
0 upgraded, 1 newly installed, 0 to remove and 126 not upgraded.
Need to get 23.0 kB of archives.
After this operation, 82.9 kB of additional disk space will be used.
Err http://ftp.debian.org/debian/ jessie/main picocom armel 1.7-1
  Temporary failure resolving 'ftp.debian.org'
E: Failed to fetch http://ftp.debian.org/debian/pool/main/p/picocom/picocom_1.7-1_armel.deb  Temporary failure resolving 'ftp.debian.org'

E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
robot@ev3dev:~$ picocom
-bash: picocom: command not found
robot@ev3dev:~$

``` 
Are there any solutions to my problem? Thanks!
